Explicitly demarcate platform support

notgull opened the pull request:
As discussed in today's meeting, this commit creates two tiers of
support. Tier 1, which is what we actively test, and Tier 2, what we
hope compiles.

MarijnS95 commented on FEATURES.md:
```diff
@@ -230,6 +230,56 @@ Changes in the API that have been agreed upon but aren't implemented across all
+|64-bit PowerPC Linux with glibc     |`powerpc64le-unknown-linux-gnu`     |X11, Wayland   |
+|32-Bit ARM Linux with glibc         |`armv5te-unknown-linux-gnueabi`     |X11, Wayland   |
+|64-Bit Linux on IBM Supercomputers  |`s390x-unknown-linux-gnu`           |X11, Wayland   |
+|32-bit ARM Android                  |`arm-linux-androideabi`             |X11, Wayland   |
```
```suggestion
|32-bit ARM Android                  |`arm-linux-androideabi`             |Android        |
```

kchibisov commented:
Should probably put it in a separate file, since we'll remove the `FEATURES.md` and then probably should also include it in the docs somehow?

notgull commented:
> Should probably put it in a separate file, since we'll remove the `FEATURES.md` and then probably should also include it in the docs somehow?

Done and done

notgull commented:
For the initial draft here I've gone with more platforms than we probably need. I think all the tier 1's are necessary, since we actively test them in CI and they directly correspond to the canonical versions of the `winit` backend's platforms.

For the tier 2's:

- ARM windows is becoming more popular, so we should make sure our bases are covered there.
- Do we want to support Windows 7? I added it here for discussion. I think we should, since lots of computers (especially old ones) still run Win7.
- Alpine Linux is sufficiently popular that I think we should be building `musl`. Same with FreeBSD. Not sure about NetBSD but issues have been filed on it in the past so I've added it.
- Other architecture Linux is fair game. I think we should be building ARM, I can take or leave the rest.

madsmtm commented on src/lib.rs:
```diff
@@ -147,6 +147,59 @@
 //! See the [`platform`] module for documentation on platform-specific cargo
 //! features.
 //!
+//! # Platform/Architecture Support
```
Either `#![doc = include_str!("PLATFORM_SUPPORT.md")]`, or to my preference, just avoid the `PLATFORM_SUPPORT.md` file altogether.

madsmtm commented on src/lib.rs:
```diff
@@ -147,6 +147,59 @@
+//! |64-Bit x86 Redox OS            |`x86_64-unknown-redox`              |Orbital        |
+//! |64-bit x64 macOS               |`x86_64-apple-darwin`               |AppKit         |
+//! |64-bit ARM macOS               |`aarch64-apple-darwin`              |AppKit         |
+//! |64-bit x64 iOS                 |`x86_64-apple-ios`                  |UIKit          |
+//! |64-bit ARM iOS                 |`aarch64-apple-ios`                 |UIKit          |
```
So, honestly, I feel like the statement "guaranteed to work" is too strong for the iOS backend. It doesn't really work properly (yet).

I'd be fine with calling it tier 2, and then I'll promote it myself once I feel like it's up there with the giants ;)

Same argument goes for Android, though it's not my call to make there.

madsmtm replied:
Also, `aarch64-apple-ios-macabi` is also tested regularly (more regularly, actually).

madsmtm reviewed (approved):
Approved once the data is deduplicated somehow (see first comment)

madsmtm commented on src/lib.rs:
```diff
@@ -147,6 +147,60 @@
+//! |64-bit x86 Android                  |`x86_64-linux-android`              |Android        |
+//! |64-bit x64 iOS                      |`x86_64-apple-ios`                  |UIKit          |
+//! |64-bit ARM iOS                      |`aarch64-apple-ios`                 |UIKit          |
+//! |64-bit ARM iOS (macOS ABI)          |`aarch64-apple-ios-macabi`          |UIKit          |
```
Nit:

```suggestion
//! |64-bit ARM Mac Catalyst             |`aarch64-apple-ios-macabi`          |UIKit          |
```

jackpot51 commented on src/lib.rs:
```diff
@@ -147,6 +147,60 @@
+//! |32-Bit x86 Linux with glibc    |`i686-unknown-linux-gnu`            |X11, Wayland   |
+//! |64-Bit x86 Linux with glibc    |`x86_64-unknown-linux-gnu`          |X11, Wayland   |
+//! |64-Bit ARM Android             |`aarch64-linux-android`             |Android        |
+//! |64-Bit x86 Redox OS            |`x86_64-unknown-redox`              |Orbital        |
```
I also test i686-unknown-redox and aarch64-unknown-redox regularly and they are functional. Should they be either Tier 1 or Tier 2?

notgull replied:
The main idea behind this list is to determine which platforms should be tested in CI and which platforms should just be `cargo check`'d. Still, I'll take your word for it and add them to Tier 1.

daxpedda commented on src/lib.rs:
```diff
@@ -147,6 +147,62 @@
+//! |64-Bit ARM Redox OS            |`aarch64-unknown-redox`             |Orbital        |
+//! |64-bit x64 macOS               |`x86_64-apple-darwin`               |AppKit         |
+//! |64-bit ARM macOS               |`aarch64-apple-darwin`              |AppKit         |
+//! |32-bit WASM web browser        |`wasm32-unknown-unknown`            |DOM            |
```
```suggestion
//! |32-bit Wasm web browser        |`wasm32-unknown-unknown`            |DOM            |
```

daxpedda commented on src/lib.rs:
```diff
@@ -147,6 +147,62 @@
+//! |64-Bit ARM Redox OS            |`aarch64-unknown-redox`             |Orbital        |
+//! |64-bit x64 macOS               |`x86_64-apple-darwin`               |AppKit         |
+//! |64-bit ARM macOS               |`aarch64-apple-darwin`              |AppKit         |
+//! |32-bit WASM web browser        |`wasm32-unknown-unknown`            |DOM            |
```
```suggestion
//! |32-bit WASM web browser        |`wasm32-unknown-unknown`            |               |
```
I think its best to leave this empty ... "DOM" isn't appropriate here and I wouldn't know what other term to use in this case.

daxpedda replied:
On second thought, I think this is the most appropriate here:
```suggestion
//! |32-bit WASM web browser        |`wasm32-unknown-unknown`            |`wasm-bindgen`|
```

daxpedda commented on src/lib.rs:
```diff
@@ -147,6 +147,62 @@
+//!
+//! - Tier 1 is **guaranteed to work**. Targets in this tier are actively tested both in CI and by
+//!   maintainers.
+//! - Tier 2 is **guaranteed to build**. Code compilation is tested in CI, but deeper testing is not
+//!   done.
```
I'm a bit confused, we don't actually test compilation of all these tier 2 targets in CI.
In the OP you said "we hope" compiles, which might be more appropriate here, but is probably very weird to add here.

notgull replied:
The goal is to start doing that in CI after we finalize this list.

daxpedda replied:
Sure!

daxpedda reviewed (changes requested):
I think this requires a tier 3?
E.g. tier 2 is actually tested in CI, tier 3 is not but we believe it should work?
Unless we want to actually just go ahead and test them in CI (which I would be in favor of), then we should do this right here.

Or am I missing something?

daxpedda reviewed (approved):
I went ahead and adjusted the wording around Web as I suggested.
